bundles/org.openhab.binding.radiothermostat/src/main/java/org/openhab/binding/radiothermostat/internal/discovery/RadioThermostatDiscoveryService.java

```diff
@@ -0,0 +1,301 @@
+                    if (ni.isUp() && ni.supportsMulticast() && !ni.isLoopback()) {
+                        sendDiscoveryBroacast(ni);
+                    }
+                }
+            } catch (IOException e) {
+                logger.debug("Error discoverying devices", e);
+            }
+        }
+    }
+
+    /**
+     * Broadcasts a SSDP discovery message into the network to find provided
+     * services.
+     *
+     * @return The Socket the answers will arrive at.
+     * @throws UnknownHostException
+     * @throws IOException
+     * @throws SocketException
+     * @throws UnsupportedEncodingException
+     */
+    private void sendDiscoveryBroacast(NetworkInterface ni)
+            throws UnknownHostException, SocketException, UnsupportedEncodingException {
+        InetAddress m = InetAddress.getByName("239.255.255.250");
+        final int port = 1900;
+        logger.debug("Sending discovery broadcast");
+        try {
+            Enumeration<InetAddress> addrs = ni.getInetAddresses();
+            InetAddress a = null;
+            while (addrs.hasMoreElements()) {
+                a = addrs.nextElement();
+                if (a instanceof Inet4Address) {
+                    break;
+                } else {
+                    a = null;
+                }
+            }
+            if (a == null) {
+                logger.debug("no ipv4 address on {}", ni.getName());
+                return;
+            }
+
+            // for whatever reason, the radio thermostat responses will not be seen
+            // if we bind this socket to a particular address.
+            // this seems to be okay on linux systems, but osx apparently prefers ipv6, so this
+            // prevents responses from being received unless the ipv4 stack is given preference.
+            MulticastSocket socket = new MulticastSocket(null);
+            socket.setSoTimeout(5000);
+            socket.setReuseAddress(true);
+            // socket.setBroadcast(true);
+            socket.setNetworkInterface(ni);
+            socket.joinGroup(m);
+            logger.debug("Joined UPnP Multicast group on Interface: {}", ni.getName());
+            byte[] requestMessage = RADIOTHERMOSTAT_DISCOVERY_MESSAGE.getBytes("UTF-8");
+            DatagramPacket datagramPacket = new DatagramPacket(requestMessage, requestMessage.length, m, port);
+            socket.send(datagramPacket);
+            try {
+                // Try to ensure that joinGroup has taken effect. Without this delay, the query
+                // packet ends up going out before the group join.
+                Thread.sleep(1000);
+
+                socket.send(datagramPacket);
+
+                byte[] buf = new byte[4096];
+                DatagramPacket packet = new DatagramPacket(buf, buf.length);
+
+                try {
+                    while (!Thread.interrupted()) {
+                        socket.receive(packet);
+                        String response = new String(packet.getData());
```

Comment:
I'm not sure I understand where it should be specified.

Reply:
There are several different constructors for `String`, just use one that allows you to specify the charset like:
```suggestion
                        String response = new String(packet.getData(), StandardCharsets.UTF8);
```

Reply:
Is this an issue for the other bindings that do discovery as well? It seems like only Feican specifies it?

```
root@Ratbert2:/mnt/d/openhab/openhab2-addons/bundles# egrep -Re 'new String.*packet.getData' *
org.openhab.binding.feican/src/main/java/org/openhab/binding/feican/internal/FeicanDiscoveryService.java:                    new String(packet.getData(), 0, packet.getLength() - 1, StandardCharsets.UTF_8));
org.openhab.binding.globalcache/src/main/java/org/openhab/binding/globalcache/internal/discovery/MulticastListener.java:        String beacon = (new String(packet.getData())).trim();
org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/discovery/LutronMcastBridgeDiscoveryService.java:            String data = new String(packet.getData(), packet.getOffset(), packet.getLength(),
org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/MieleBridgeHandler.java:                                String event = new String(packet.getData());
org.openhab.binding.milight/src/main/java/org/openhab/binding/milight/internal/handler/BridgeV3Handler.java:                final String received = new String(packet.getData());
org.openhab.binding.radiothermostat/src/main/java/org/openhab/binding/radiothermostat/internal/discovery/RadioThermostatDiscoveryService.java:                        String response = new String(packet.getData());
```

Reply:
I think that is because you are searching for `String(packet.getData()` which I'm pretty sure a lot bindings don't use that exact text

Reply:
But yes, this would be an issue with other bindings as well. But the default charset is usually ok since it is US_ASCII or UTF-8, it only becomes a problem when running on machines with vastly different locales like arabic or chinese.

Reply:
@bobadair, one of them is Lutron.

Reply:
> one of them is Lutron

Thanks Bill. I checked that file in the Lutron binding, and it actually does specify the US_ASCII charset. It just does it on the next line. :-)

```
            String data = new String(packet.getData(), packet.getOffset(), packet.getLength(),
                    StandardCharsets.US_ASCII);
```